lib/seattleflu/db/cli/command/preprocess.py

```diff
@@ -0,0 +1,288 @@
+
+    nwh_df = load_nwh_manifest_data(uw_nwh_file)
+    nwh_df = nwh_df.rename(columns={'Barcode ID (Sample ID)': 'Barcode ID',
+                                    'Collection Date (per tube)': 'Collection date'})
+
+    hmc_df = load_hmc_manifest_data(hmc_sch_file)
+
+    return df, uw_df, nwh_df, hmc_df
+
+def load_uw_metadata(uw_filename: str) -> pd.DataFrame:
+    if uw_filename.endswith('.csv'):
+        df = pd.read_csv(uw_filename)
+    else:
+        df = pd.read_excel(uw_filename, sheet_name='pts')
+    return df
+
+def load_uw_manifest_data(uw_nwh_file: str) -> pd.DataFrame:
+    return pd.read_excel(uw_nwh_file, sheet_name='UWMC', keep_default_na=False,
+                         usecols=['Barcode ID (Sample ID)', 'MRN', 
+                                  'Collection Date', 'Accession'],
+                         na_values=['NA', ''])
+
+def load_nwh_manifest_data(uw_nwh_file: str) -> pd.DataFrame:
+    return pd.read_excel(uw_nwh_file,sheet_name='NWH',keep_default_na=False,
+                         usecols=['Barcode ID (Sample ID)', 'MRN', 
+                                  'Collection Date (per tube)', 'Accession'],
+                         na_values=['NA', ''])
+
+def load_hmc_manifest_data(hmc_sch_file: str) -> pd.DataFrame:
+    return pd.read_excel(hmc_sch_file,sheet_name='HMC',keep_default_na=False,
+                         usecols=['Barcode ID', 'MRN', 'Accession', 
+                                  'Collection date'], 
+                         na_values=['NA', ''])
+
+def create_unique_identifier(df: pd.DataFrame): 
+    """Generate a unique identifier for each encounter and drop duplicates"""
+    df['identifier'] = (df['labMRN'] + df['LabAccNum'] + \
+                        df['Collection date'].astype(str)
+                        ).str.lower()
+    return df.drop_duplicates(subset="identifier") 
+
+
+def standardize_identifiers(df: pd.DataFrame) -> pd.DataFrame:
+    """Convert all to lower case for matching purposes"""
+
+    df['MRN'] = df['MRN'].str.lower()
+    df['Accession'] = df['Accession'].str.lower()
+    return df
+
+
+def drop_columns(columns_to_keep: list, df: pd.DataFrame):
+    """
+    Given a list of column names and a pandas DataFrame,
+    drop all columns in Dataframe that are not in list. 
+    """
+    columns = list(df.columns)
+    columns_to_drop = [col for col in columns if col not in columns_to_keep]
+    return df.drop(columns=columns_to_drop, axis=1)
+
+
+def missing_barcode(df: pd.DataFrame): 
+    """ TODO """
+    missing_barcodes = df.loc[df['barcode'].isnull()].copy()
+    missing_barcodes['problem'] = 'Missing barcode'
+    return generate_problem_barcode_columns(missing_barcodes)
+
+
+def generate_problem_barcode_columns(df: pd.DataFrame):
+    """ TODO Should we just carry over MRN, Accession and Date instead of 
+        doing it this way? 
+    """ 
+    df.loc[:, 'MRN'] = df['identifier'].str[:8]
+    df.loc[:, 'Accession'] = df['identifier'].str[8:]
```

Comment:
To avoid this, we can simply drop our MRN and Accession columns after we grab our duplicates. To quote Mike Famulare on Tuesday, some code may have "degenerated a bit in the last 24 hours". :)